doc/debugfs.reiser4.8
.\"						Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH debugfs.reiser4 8 "02 Oct, 2002" reiser4progs "reiser4progs manual"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
debugfs.reiser4 \- the program for debugging reiser4 filesystem.
.SH SYNOPSIS
.B debugfs.reiser4
[ options ] FILE
.SH DESCRIPTION
.B debugfs.reiser4
is reiser4 filesystem debug program. You can disscover the internal
reiser4 filesystem structures by using it.
.SH COMMON OPTIONS
.TP
.B -V, --version
prints program version.
.TP
.B -?, -h, --help
prints program help.
.TP
.B -y, --yes
assumes an answer 'yes' to all questions.
.TP
.B -f, --force
forces debugfs to use whole disk, not block device or mounted partition.
.TP
.B -c, --cache N
sets tree cache node number to passed value. This affects very much
behavior of libreiser4. It affects speed, tree allocation, etc.
.SH BROWSING OPTIONS
.TP
.B -k, --cat
browses passed file like standard cat and ls programs.
.SH PRINT OPTIONS
.TP
.B -t, --print-tree
prints the internal tree.
.TP
.B -b, --print-block N
prints the block associated with the passed block number.
.TP
.B -n, --print-nodes FILE
prints all nodes that the passed file lies in.
.TP
.B -i, --print-file
prints all items that the passed file consists of.
.TP
.B -s, --print-super
prints the both super blocks: master super block and format specific one.
.TP
.B -j, --print-journal
prints the journal with not yet commited transactions (if any).
.TP
.B -a, --print-alloc
prints the block allocator data.
.TP
.B -d, --print-oid
prints the oid allocator data.
.SH METADATA OPTIONS
.TP
.B -P, --pack-metadata
fetches filesystem metadata and writes them to the standard output.
.TP
.B -U, --unpack-metadata
reads filesystem metadata stream from the stdandard input and constructs 
a new filesystem based on the metadata. debugfs.reiser4 --pack-metadata <FS1> | 
debugfs.reiser4 --unpack-metadata <FS2> and then debugfs.reiser4 --pack-metadata 
<FS2> produces a stream equivalent to the first one.
.SH PLUGIN OPTIONS
.TP
.B -p, --print-profile
prints the plugin profile. This is the set of default plugins used for all
parts of a filesystem -- format, nodes, files, directories, hashes, etc. If
--override is specified, then prints modified plugins.
.TP
.B -l, --print-plugins
prints all plugins libreiser4 knows about.
.TP
.B -o, --override TYPE=PLUGIN, ...
overrides the default plugin of the type "TYPE" by the plugin "PLUGIN"
in the plugin profile.
.SH Examples:
.sp 1
debugfs.reiser4 -o nodeptr=nodeptr41,hash=rupasov_hash /dev/hda2
.RS
.SH REPORTING BUGS
Report bugs to <reiserfs-devel@vger.kernel.org>
.SH SEE ALSO
.BR measurefs.reiser4(8),
.BR mkfs.reiser4(8),
.BR fsck.reiser4(8)
.SH AUTHOR
This manual page was written by Yury Umanets <umka@namesys.com>


ChangeLog
This file is part of reiser4progs package

Copyright (C) 2001, 2002, 2003, 2004 by Hans Reiser, licensing governed by
reiser4progs/COPYING.

This file may be modified and/or distributed without restriction. This is
not an invitation to misrepresent the history of reiser4progs.

2015-08-14
	1.1.0 version
2015-08-14
	Format release 4.0.1;
	Support checksum for formatted nodes;
	NODE41 format with 32-bit field for checksums;
2014-06-28
	1.0.9 version
2014-06-28
	discard option to mkfs -- discard device at filesystem creation
	time (for SSD drives)

	Fixup in librepair/twig_scan.c: cb_item_region_check().
2013-05-03
	1.0.8 version
2013-05-03
	Fixups in fsck.reiser4: repair a special kind of corruptions
	appeared due to software bugs in earlier kernels.

	mkfs.reiser4 now enables intelligent compression with lzo1
	algorithm by default. If you want to create reiser4 without
	compression, then override the default plugin set with the
	"-o create=reg40" option.
2008-01-09
	1.0.7 version.
2008-01-09
	Alias -p "preen" to -a "auto" for fsck.
	Bugfixes:
	- fixed bugs in CryptoCompression object plugin (incorrect handling
	of corrupted compressed bodies).
	- suppress messages about wrong bytes for CryptoCompression
	objects	as in most cases it is because kernel doesn't support
	i_blocks and i_bytes for such objects because of performance
	issues. If wrong bytes is the result of fatal corruption, then
	fsck will detect and report it.
2006-02-22
	1.0.6 version.
2006-01-13
	Version control data is added into format40 disk format plugin.
	Version control data is added into backup blocks.
	Recover filesystems of both old and new formats - temporary.
	UPDATE_BACKUP_IS_NEEDED is added -- handle correctly cases when 
	disk format is updated and backup blocks are not; fsck correctly.
2005-12-09
	Heir plugin SET (HSET) is added.
2005-11-21
	Bugfixes:
	- a lot of fixes in compression tail item plugin;
	- a lot of fixes in CryptoCompression object plugin;
	- a lot of fixes in common object plugin repair code;
	- in spec file.
2005-11-08
	A lot of speedups.
	A lot of bugfixes.
	Demos/busy program update: crypto compression test is added.
2005-10-31
	Bugfixes in truncate, clubber and check_struct methods of all object
	plugins.
2005-10-28
	Compression plugin is added. 
	CryptoCompression plugin is adjusted to handle compressed files.
2005-10-14
	Bugfixes:
	- check the path given on a symlink creation;
	- check the possible file position overflow;
	- handle the cluster size of ctail items correctly.
2005-09-31
	Crypto-Compression Object plugins is added.
	Fix the object creation: inherit always from the parent.
	Fix the pset storage policy.
2005-08-31
	Compression tail item (ctail) plugin is added.
2005-08-16
	The full support for Crypto-Compression plugins is added into PluginSet
	and libreiser4 profile.
	Crypto SD extention is edded.
2005-08-12
	1.0.5 version.
2005-08-10
	Installation fix.
	Return value fixes.
	Backup block handling fixes.
2005-06-28
	Ability to link against libaal placed at arbitrary location is fixed.
	Unlink dir fix.
	Stat40 check_struct fix.
	Man pages for not ready utils are removed.
	Handle new (crc related) fields in plugin set correctly.
	Flush fs metadata fix (interrupt related).
	Do not flush dirty nodes if fsck fails.
	Recalculate bitmap checksums at the end of fsck if they are wrong.
	Do not update bitmaps after every fsck pass if no metadata found.
2005-03-23
	nomkid option to fsck -- ability to recover reiser4 w/out mkid check.
2005-03-21
	Large file handling fixes.
2005-03-16
	Some casting fixes.
2005-02-21
	Reg40 file body convertion fixes.
2005-02-20
	The 1.0.4 version.
2005-02-19
	A bugfix in file body convertion code.
2005-02-17
	Enable libminimal by default.
	Do not configure empty utils(resizer, cpfs).
2005-01-30
	Some bug fixes in tree balancing, fs check, syncing code.
2005-01-25
	Some bug fixes in tree balancing code.
2005-01-20
	Bug fixes:
	- in backup layout traverse code;
	- in insert extent item-by-item code;
2005-01-17
	A way to recover fs metadata independently of the mkfs id is added.
	Bug fixes:
	- in repair code then a node has a pointer to itself;
	- in not aligned access to different object on-disk data.
2004-12-23
	Bug fixes:
	- in initializing object id fields in SB;
	- in free blocks accounting;
	- in journal unpack code.
2004-12-12
	Some bugfixes:
	- in object/extent truncate code are fixed.
	- in backup handling code is fixed.
	- sware about libaal version correctly.
2004-12-04
	The version is 1.0.3
2004-12-01
	The Pack/unpack journal code is written.
	Cleanups in program options.
2004-11-25
	Build either all or none progs statically.
	The recover from unused space code is added. Could be used as undelete.
2004-11-24
	The backup code is written.
2004-10-29
	A bunch of bugs is fixed.
2004-10-21
	Object detach/unlink methods are fixed.
	Fix the handling of special entries ('.', '..') in directory objects.
	The version is 1.0.2.
2004-10-15
	File body write code (namely extent40 write) is rewritten
	and does work now.
2004-10-12
	Make allocator CRC endian safe
2004-10-07
	A bunch of bugs is fixed in the balancing code
2004-09-22
	Change "stand alone" to "minimal footprint"
2004-09-21
	Get rid of key collision problem is tree_scan
2004-09-16
	sd_bytes fix for files larger then 4G.
2004-09-14
	Make ldconfig to run automatically on 'make install'.
2004-09-13
	Make r5 hash working correctly with extended ASCII characters.
2004-08-24
	Some bugs are fixed:
	- in tree_scan method;
	- in reading extent items;
	- in man pages.
	Some optimisations for stand-alone are added.
	Version is 1.0.1.
2004-08-09
	1.0.0 version is released.


progs/debugfs/debugfs.c
/* Copyright (C) 2001-2005 by Hans Reiser, licensing governed by
   reiser4progs/COPYING.
   
   debugfs.c -- program for debugging reiser4 filesystem. */

#include <fcntl.h>
#include <getopt.h>
#include <errno.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>

#include "debugfs.h"
#include "aux/aux.h"

/* Prints debugfs options */
static void debugfs_print_usage(char *name) {
	fprintf(stderr, "Usage: %s [ options ] FILE\n", name);
    
	fprintf(stderr, 
		"Browsing options:\n"
		"  -k, --cat FILE                browses passed file like standard\n"
		"                                cat and ls programs.\n"
		"Print options:\n"
		"  -s, --print-super             prints the both super blocks.\n"
		"  -t, --print-tree              prints the whole tree.\n"
		"  -j, --print-journal           prints journal.\n"
		"  -d, --print-oid               prints oid allocator data.\n"
		"  -a, --print-alloc             prints block allocator data.\n"
		"  -b, --print-block N           prints block by its number.\n"
		"  -n, --print-nodes FILE        prints all nodes file lies in.\n"
		"  -i, --print-file FILE         prints all items specified file\n"
		"                                consists of.\n"
		"Metadata options:\n"
		"  -P, --pack-metadata           fetches filesystem metadata and writes it\n"
		"                                to standard output.\n"
		"  -U, --unpack-metadata         uses metadata stream from stdandard input\n"
		"                                to construct filesystem by it.\n"
		"Space options:\n"
		"  -O, --occupied-blocks         works with occupied blocks only(default).\n"
		"  -F, --free-blocks             works with free blocks only.\n"
		"  -W, --whole-partition         works with the whole partition.\n"
		"  -B, --bitmap                  works with blocks marked in the bitmap only.\n"
		"Plugins options:\n"
		"  -p, --print-profile           prints the plugin profile.\n"
		"  -l, --print-plugins           prints known plugins.\n"
	        "  -o, --override TYPE=PLUGIN    overrides the default plugin of the type\n"
	        "                                \"TYPE\" by the plugin \"PLUGIN\" in the\n"
		"                                profile.\n"
		"Common options:\n"
		"  -?, -h, --help                prints program usage.\n"
		"  -V, --version                 prints current version.\n"
		"  -f, --force                   makes debugfs to use whole disk, not\n"
		"  -y, --yes                     assumes an answer 'yes' to all questions.\n"
		"                                block device or mounted partition.\n"
		"  -c, --cache N                 number of nodes in tree buffer cache\n");
}

/* Initializes exception streams used by debugfs */
static void debugfs_init(void) {
	int ex;

	/* Setting up exception streams. */
	for (ex = 0; ex < EXCEPTION_TYPE_LAST; ex++)
		misc_exception_set_stream(ex, stderr);
	
	/* All steams should go to stderr, as in/out are used for pack/unpack 
	   and other needs */
	misc_exception_set_stream(EXCEPTION_TYPE_MESSAGE, stderr);
	misc_exception_set_stream(EXCEPTION_TYPE_INFO, stderr);
	misc_exception_set_stream(EXCEPTION_TYPE_FSCK, NULL);
}

typedef struct debugfs_backup_hint {
	uint64_t count;
	blk_t *blk;
	aal_gauge_t *gauge;
} debugfs_backup_hint_t;

enum {
	NF_DONE = NF_LAST
};

extern void cb_gauge_tree_percent(aal_gauge_t *gauge);

int main(int argc, char *argv[]) {
	int c;

	uint32_t cache;
	struct stat st;
	char *host_dev;

	uint32_t print_flags = 0;
	uint32_t behav_flags = 0;
	uint32_t space_flags = 0;

	char override[4096];
	char *cat_filename = NULL;
	char *print_filename = NULL;
    
	aal_device_t *device;
	reiser4_fs_t *fs = NULL;
	blk_t blocknr = 0;

	FILE *file = NULL;
	char *bm_file = NULL;
	reiser4_bitmap_t *bitmap = NULL;
	
	static struct option long_options[] = {
		{"version", no_argument, NULL, 'V'},
		{"help", no_argument, NULL, 'h'},
		{"force", no_argument, NULL, 'f'},
		{"yes", no_argument, NULL, 'y'},
		{"cat", required_argument, NULL, 'k'},
		{"print-tree", no_argument, NULL, 't'},
		{"print-journal", no_argument, NULL, 'j'},
		{"print-super", no_argument, NULL, 's'},
		{"print-alloc", no_argument, NULL, 'a'},
		{"print-oid", no_argument, NULL, 'd'},
		{"print-block", required_argument, NULL, 'b'},
		{"print-nodes", required_argument, NULL, 'n'},
		{"print-file", required_argument, NULL, 'i'},
		{"pack-metadata", no_argument, NULL, 'P'},
		{"unpack-metadata", no_argument, NULL, 'U'},
		{"print-profile", no_argument, NULL, 'p'},
		{"print-plugins", no_argument, NULL, 'l'},
		{"override", required_argument, NULL, 'o'},
		{"occupied-blocks", no_argument, NULL, 'O'},
		{"free-blocks", no_argument, NULL, 'F'},
		{"bitmap", required_argument, NULL, 'B'},
		{"whole-partition", no_argument, NULL, 'W'},
		{"cache", required_argument, NULL, 'c'},
		{0, 0, 0, 0}
	};

	debugfs_init();
	memset(override, 0, sizeof(override));
	
	if (argc < 2) {
		debugfs_print_usage(argv[0]);
		return USER_ERROR;
	}
    
	/* Parsing parameters */    
	while ((c = getopt_long(argc, argv, "hVyftb:djk:n:i:o:plsaPUOFWB:c:?",
				long_options, (int *)0)) != EOF) 
	{
		switch (c) {
		case 'h':
		case '?':
			debugfs_print_usage(argv[0]);
			return NO_ERROR;
		case 'V':
			misc_print_banner(argv[0]);
			return NO_ERROR;
		case 'd':
			print_flags |= PF_OID;
			break;
		case 'a':
			print_flags |= PF_ALLOC;
			break;
		case 's':
			print_flags |= PF_SUPER;
			break;
		case 'j':
			print_flags |= PF_JOURNAL;
			break;
		case 't':
			print_flags |= PF_TREE;
			break;
		case 'b': {
			print_flags |= PF_BLOCK;

			/* Parsing block number */
			if ((blocknr = misc_str2long(optarg, 10)) == INVAL_DIG) {
				aal_error("Invalid block number (%s).", optarg);
				return USER_ERROR;
			}
			
			break;
		}
		case 'n':
			print_flags |= PF_NODES;
			print_filename = optarg;
			break;
		case 'i':
			print_flags |= PF_ITEMS;
			print_filename = optarg;
			break;
		case 'k':
			behav_flags |= BF_CAT;
			cat_filename = optarg;
			break;
		case 'c':
			if ((cache = misc_str2long(optarg, 10)) == INVAL_DIG) {
				aal_error("Invalid cache value specified (%s).",
					  optarg);
				return USER_ERROR;
			}

			misc_mpressure_setup(cache);
			break;
		case 'f':
			behav_flags |= BF_FORCE;
			break;
		case 'y':
			behav_flags |= BF_YES;
			break;
		case 'P':
			behav_flags |= BF_PACK_META;
			break;
		case 'U':
			behav_flags |= BF_UNPACK_META;
			break;
		case 'l':
			behav_flags |= BF_SHOW_PLUG;
			break;
		case 'p':
			behav_flags |= BF_SHOW_PARM;
			break;
		case 'o':
			aal_strncat(override, optarg, aal_strlen(optarg));
			aal_strncat(override, ",", 1);
			break;
		case 'O':
			space_flags = 0;
			break;
		case 'F':
			space_flags |= SF_FREE;
			break;
		case 'W':
			space_flags |= SF_WHOLE;
			break;
		case 'B':
			bm_file = optarg;
			break;
		}
	}
    
	if (!(behav_flags & BF_YES))
		misc_print_banner(argv[0]);

	if (libreiser4_init()) {
		aal_error("Can't initialize libreiser4.");
		goto error;
	}

	/* Overriding default params by passed values. This should be done after
	   libreiser4 is initialized. */
	if (aal_strlen(override) > 0) {
		override[aal_strlen(override) - 1] = '\0';
		
		if (!(behav_flags & BF_YES)) {
			aal_mess("Overriding the plugin profile by \"%s\".", 
				 override);
		}
		
		if (misc_profile_override(override))
			goto error_free_libreiser4;
	}
	
	if (behav_flags & BF_SHOW_PARM)
		misc_profile_print();

	if (behav_flags & BF_SHOW_PLUG)
		misc_plugins_print();

	if (optind >= argc)
		goto error_free_libreiser4;
		
	host_dev = argv[optind];
    
	if (stat(host_dev, &st) == -1) {
		aal_error("Can't stat %s. %s.", host_dev, strerror(errno));
		goto error_free_libreiser4;
	}
	
	/* Checking if passed partition is mounted */
	if (misc_dev_mounted(host_dev) == MF_RW && !(behav_flags & BF_FORCE)) {
		aal_error("Device %s is mounted 'rw' at the moment. "
			  "Use -f to force over.", host_dev);
		goto error_free_libreiser4;
	}

	/* Opening device with file_ops and default blocksize */
	if (!(device = aal_device_open(&file_ops, host_dev, 512,
				       behav_flags & BF_UNPACK_META ?
				       O_RDWR : O_RDONLY)))
	{
		aal_error("Can't open %s. %s.", host_dev, strerror(errno));
		goto error_free_libreiser4;
	}

			
	if (behav_flags & BF_UNPACK_META) {
		aal_stream_t stream;
		char buf[256];
		int i;
		
		/* Prepare the bitmap if needed. */
		if (bm_file) {
			if ((file = fopen(bm_file, "w+")) == NULL) {
				aal_fatal("Can't open the bitmap "
					  "file (%s).", bm_file);
				goto error_free_device;
			}

			if (!(bitmap = reiser4_bitmap_create(0))) {
				aal_error("Can't allocate a bitmap.");
				goto error_close_file;
			}
		}
		
		aal_stream_init(&stream, stdin, &file_stream);
		aal_stream_read(&stream, buf, 4);
		
		if (aal_strncmp(buf, VERSION_PACK_SIGN, 4)) {
			aal_error("The metadata were packed with the "
				  "reiser4progs version <= 1.0.2.");
			goto error_free_bitmap;
		}
		
		i = 0;
		while (1) {
			if (aal_stream_read(&stream, buf + i, 1) != 1) {
				aal_error("Can't read from the stream. Is it over?");
				goto error_free_bitmap;
			}
			
			if (buf[i] == '\0' || i == 255)
				break;
			i++;
		}

		if (i == 255) {
			aal_fatal("Can't detect the reiser4progs version "
				  "the metadata were packed with.");
			goto error_free_bitmap;
		} else {
			aal_info("The metadata were packed with the "
				 "reiser4progs %s.", buf);
		}
		
		if (!(fs = repair_fs_unpack(device, bitmap, &stream))) {
			aal_error("Can't unpack filesystem.");
			goto error_free_bitmap;
		}

		aal_stream_fini(&stream);

		if (reiser4_fs_sync(fs)) {
			aal_error("Can't save unpacked filesystem.");
			goto error_free_bitmap;
		}

		/* Save the bitmap. */
		if (bitmap) {
			aal_stream_init(&stream, file, &file_stream);

			if (reiser4_bitmap_pack(bitmap, &stream)) {
				aal_error("Can't pack the bitmap of "
					  "unpacked blocks.");
				goto error_free_bitmap;
			}

			aal_stream_fini(&stream);
			reiser4_bitmap_close(bitmap);
			fclose(file);
			bitmap = NULL;
			file = NULL;
		}

		goto done;
	} else {
		/* Open file system on the device */
		if (!(fs = reiser4_fs_open(device, 0))) {
			aal_error("Can't open reiser4 on %s", host_dev);
			goto error_free_device;
		}

		/* Opening the journal */
		if (!(fs->journal = reiser4_journal_open(fs, device))) {
			aal_error("Can't open journal on %s", host_dev);
			goto error_free_fs;
		}
	}
	
	if (behav_flags & BF_PACK_META /* || print disk blocks */) {
		uint64_t len = reiser4_format_get_len(fs->format);
		aal_stream_t stream;
		
		if (bm_file) {
			/* Read the bitmap from the file. */
			if ((file = fopen(bm_file, "r")) == NULL) {
				aal_fatal("Can't open the bitmap "
					  "file (%s).", bm_file);
				goto error_free_fs;
			}
		
			aal_stream_init(&stream, file, &file_stream);

			if (!(bitmap = reiser4_bitmap_unpack(&stream))) {
				aal_error("Can't unpack the bitmap of "
					  "packed blocks.");
				goto error_close_file;
			}

			aal_stream_fini(&stream);
			fclose(file);
			file = NULL;
		} else {
			if (!(bitmap = reiser4_bitmap_create(len))) {
				aal_error("Can't allocate a bitmap.");
				goto error_free_fs;
			}

			/* Used || free blocks - extract allocator to bitmap. */
			if (space_flags == 0 || space_flags & SF_FREE) {
				if (reiser4_alloc_extract(fs->alloc, bitmap)) {
					aal_error("Can't extract the space "
						  "allocator data to bitmap.");
					goto error_free_bitmap;
				}
			}

			/* Whole partition || free blocks - invert bitmap. */
			if (space_flags & SF_WHOLE || space_flags & SF_FREE) {
				reiser4_bitmap_invert(bitmap);
			}
		}
	}

	/* In the case no print flags was specified, debugfs will print super
	   blocks by defaut. */
	if (print_flags == 0 && (behav_flags & ~(BF_FORCE | BF_YES)) == 0)
		print_flags = PF_SUPER;

	/* Handling print options */
	if ((behav_flags & BF_CAT)) {
		if (debugfs_browse(fs, cat_filename))
			goto error_free_bitmap;
	}
	
	if (print_flags & PF_SUPER) {
		debugfs_print_master(fs);
		debugfs_print_format(fs);
		debugfs_print_status(fs);
	}
    
	if (print_flags & PF_OID)
		debugfs_print_oid(fs);
    
	if (print_flags & PF_ALLOC)
		debugfs_print_alloc(fs);
    
	if (print_flags & PF_JOURNAL)
		debugfs_print_journal(fs);
    
	if (print_flags & PF_TREE)
		debugfs_print_tree(fs);

	if (print_flags & PF_BLOCK) {
		if (debugfs_print_block(fs, blocknr))
			goto error_free_bitmap;
	}
    
	if (print_flags & PF_NODES || print_flags & PF_ITEMS) {
		if (debugfs_print_file(fs, print_filename, print_flags))
			goto error_free_bitmap;
	}

	if (behav_flags & BF_PACK_META) {
		aal_stream_t stream;

		aal_stream_init(&stream, stdout, &file_stream);
		aal_stream_write(&stream, VERSION_PACK_SIGN, 4);
		aal_stream_write(&stream, VERSION, sizeof(VERSION));
		
		if (repair_fs_pack(fs, bitmap, &stream)) {
			aal_error("Can't pack filesystem.");
			goto error_free_bitmap;
		}
		
		aal_stream_fini(&stream);
		reiser4_bitmap_close(bitmap);
		bitmap = NULL;
	}
	
 done:
	/* Closing filesystem itself */
	reiser4_fs_close(fs);

	/* Closing device */
	aal_device_close(device);
    
	/* Deinitializing libreiser4. At the moment only plugins are unloading
	   while doing this. */
	libreiser4_fini();
	return NO_ERROR;

 error_free_bitmap:
	if (bitmap) {
		reiser4_bitmap_close(bitmap);
	}
 error_close_file:
	if(file) {
		fclose(file);
	}
 error_free_fs:
	if (fs) {
		reiser4_fs_close(fs);
	}
 error_free_device:
	aal_device_close(device);
 error_free_libreiser4:
	libreiser4_fini();
 error:
	return OPER_ERROR;
}

